### ISSUE TITLE
Compatibility of shared_examples with Ruby 3

### ISSUE THREAD

**Issue by WojciechKo** [rspec/rspec-core#2674]
### Subject of the issue
Compatibility of `shared_examples` with a ["Real" keyword argument](https://bugs.ruby-lang.org/issues/14183) from upcoming Ruby 3

### Your environment
* Ruby version: 2.7.0preview3
* rspec-core version: 3.9.0

### Steps to reproduce
```ruby
RSpec.describe 'ruby 2.7.0.preview3 warning' do
  shared_examples 'expects hash' do |hash|
    it { expect(hash[:input].reverse).to eq(hash[:output]) }
  end

  shared_examples 'expects keywords' do |input:, output:|
    it { expect(input.reverse).to eq(output) }
  end

  include_examples 'expects hash', input: 'afla', output: 'alfa'
  include_examples 'expects hash', { input: 'ovarb', output: 'bravo' }

  include_examples 'expects keywords', input: 'eilrahc', output: 'charlie' # Warning
  include_examples 'expects keywords', { input: 'atled', output: 'delta' } # Warning
end
```

### Expected behavior
- `charlie` example should not produce any warning
- `alfa` should not be supported to follow Ruby 3 changes (?)

### Actual behavior
Following warning is being printed for `charlie` and `delta` case
```
/home/wojciechko/.rbenv/versions/2.7.0-preview3/lib/ruby/gems/2.7.0/gems/rspec-core-3.9.0/lib/rspec/core/shared_example_group.rb:36: warning: The last argument is used as the keyword parameter
/home/wojciechko/dev/sample_spec.rb:6: warning: for method defined here
```

**Comment by JonRowe** [comment]
Work here is needed to figure out what they've changed and why this is issuing a warning, I suspect we just haven't considered keyword argument at all and they've just worked due to their hash like nature. The implementation should be agnostic to what is passed in and just pass through the arguments. (Meaning the "alfa" example will not work on ruby 3, but continue to work everywhere else).

**Comment by jaynetics** [comment]
This also affects `rspec-expectations`: https://github.com/rspec/rspec-expectations/issues/1150
and `rspec-mocks`: https://github.com/rspec/rspec-mocks/issues/1306

I was beginning to prepare PRs for these things that simply pass `**kwargs` trough all participating methods. This does fix these issues. However, in `rspec-core` that produced some other problems with example metadata being lost, which apparently passes through the same code paths. While I looked into that I noticed to my amazement that you still support ruby 1.9 and even 1.8, so that whole approach won't even work (at least without a lot of insane, RUBY_VERSION-based eval'ing).

Perhaps this is a sensible time to stop supporting these versions?

**Comment by JonRowe** [comment]
Metadata is a hash, so its not going to work treating it as keyword arguments everywhere, as we expect it to be a hash. So this warning is actually a symptom of mis-using the implementation.... So I'm kind of tempted to issue an explicit warning explaining that rather than trying to add this.

> While I looked into that I noticed to my amazement that you still support ruby 1.9 and even 1.8, so that whole approach won't even work (at least without a lot of insane, RUBY_VERSION-based eval'ing).
>
> Perhaps this is a sensible time to stop supporting these versions?

I'm afraid that won't happen until the next major version due to semver.

**Comment by pirj** [comment]
Wondering if the change to:
```diff
-      def self.include_examples(name, *args, &block)
-        find_and_eval_shared("examples", name, caller.first, *args, &block)
+      def self.include_examples(name, *args, **kwargs, &block)
+        find_and_eval_shared("examples", name, caller.first, *args, **kwargs, &block)
```
will do the job.
Still, `define_nested_shared_group_method` and `include_context` needs to be tweaked the same way.
Please correct me if I'm wrong, it seems there's no way around `eval` and adding `strict_kwargs?` to `rspec-support`.

A temporary measure around warnings can be the usage of `ruby2_keywords` macro (also needs to be wrapped, to only be called when it's available).

**Comment by pirj** [comment]
Related https://github.com/rubocop-hq/rubocop/pull/7584

**Comment by JonRowe** [comment]
Yeah I mean you should be able to remove the warnings by just accepting the hash as is...

**Comment by pirj** [comment]
Would you like to be interested in sending a pull request to fix that issue @WojciechKo ?

**Comment by JonRowe** [comment]
We fixed this in #2716 but adding further specs in #2719, closing.
